From 487b855d75f756a175cc4783110b7dd48471164a Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Sun, 3 Apr 2022 13:29:30 +0300
Subject: [PATCH 13/17] imx8mm: move mac-address selection logic to board-file

Make mac address selection explicit in the board-file to allow full
control over selection priority if multiple locations may contain MACs.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 .../imx8mm_solidrun/imx8mm_solidrun.c         | 24 +++++++++++++++++++
 drivers/net/fec_mxc.c                         |  4 ++--
 2 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/board/solidrun/imx8mm_solidrun/imx8mm_solidrun.c b/board/solidrun/imx8mm_solidrun/imx8mm_solidrun.c
index 45d8f69925..8c1ab60bfe 100644
--- a/board/solidrun/imx8mm_solidrun/imx8mm_solidrun.c
+++ b/board/solidrun/imx8mm_solidrun/imx8mm_solidrun.c
@@ -660,3 +660,27 @@ int is_recovery_key_pressing(void)
 }
 #endif /*CONFIG_ANDROID_RECOVERY*/
 #endif /*CONFIG_FSL_FASTBOOT*/
+
+/*
+ * select board mac address for given interface
+ */
+int board_get_mac(int dev_id, unsigned char *mac) {
+	char macenv[16] = {0};
+	mac[0] = 0;
+
+	/*
+	 * Note: Environment ethaddr (eth1addr, eth2addr, ...) has first priority,
+	 * therefore it should be read and returned here.
+	 * However the fec driver will write the result from this function to the environment,
+	 * causing a feedback loop.
+	 */
+
+	// fuses
+	imx_get_mac_from_fuse(dev_id, mac);
+	if(is_valid_ethaddr(mac)) {
+		printf("%s: interface %i: using mac from fuses: %02X:%02X:%02X:%02X:%02X:%02X\n", __func__, dev_id, mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+		return 0;
+	}
+
+	return -ENOENT;
+}
diff --git a/drivers/net/fec_mxc.c b/drivers/net/fec_mxc.c
index cc23aeacf5..7f76ca70c9 100644
--- a/drivers/net/fec_mxc.c
+++ b/drivers/net/fec_mxc.c
@@ -351,7 +351,7 @@ static void fec_rbd_clean(int last, struct fec_bd *prbd)
 
 static int fec_get_hwaddr(int dev_id, unsigned char *mac)
 {
-	imx_get_mac_from_fuse(dev_id, mac);
+	board_get_mac(dev_id, mac);
 	return !is_valid_ethaddr(mac);
 }
 
@@ -1121,7 +1121,7 @@ static int fec_probe(bd_t *bd, int dev_id, uint32_t base_addr,
 	edev->index = fec->dev_id;
 
 	if (fec_get_hwaddr(fec->dev_id, ethaddr) == 0) {
-		debug("got MAC%d address from fuse: %pM\n", fec->dev_id, ethaddr);
+		debug("got MAC%d address from board-file: %pM\n", fec->dev_id, ethaddr);
 		memcpy(edev->enetaddr, ethaddr, 6);
 		if (fec->dev_id)
 			sprintf(mac, "eth%daddr", fec->dev_id);
-- 
2.35.3

